service/history/workflow/context.go

```diff
@@ -916,13 +916,19 @@ func (c *ContextImpl) ReapplyEvents(
 // TODO: remove `ms` parameter again (added since it's not possible to initialize a new Context with a specific MutableState)
 func (c *ContextImpl) UpdateRegistry(ctx context.Context, ms MutableState) update.Registry {
 	if c.updateRegistry == nil {
-		if ms == nil {
-			ms = c.MutableState
+		var nsIDStr string
+		if c.MutableState != nil {
+			nsIDStr = c.MutableState.GetNamespaceEntry().ID().String()
+		} else {
+			nsIDStr = ms.GetNamespaceEntry().ID().String()
 		}
-
-		nsIDStr := ms.GetNamespaceEntry().ID().String()
 		c.updateRegistry = update.NewRegistry(
-			func() update.Store { return ms },
+			func() update.Store {
+				if c.MutableState != nil {
+					return c.MutableState
+				}
+				return ms
+			},
```

Comment:
Can you add some comments explaining what roles the two different mutable states are playing in this function?

Reply:
I updated PR description.

Reply:
I'd like this explained in code comments, since it's very unclear what's going on. Usually a passed-in non-nullable parameter would take precedence, but in any case it's unclear why we have two mutable states. If the whole situation is an unfortunate mess that we're planning to get rid of that would be helpful to say in a comment!